src/rust/engine/graph/src/lib.rs

```diff
@@ -52,22 +52,71 @@ impl RunToken {
   }
 }
 
+///
+/// A token associated with a Node that is incremented whenever its output value has (or might
+/// have) changed. When a dependent consumes a dependency at a particular generation, that
+/// generation is recorded on the consuming edge, and can later used to determine whether the
+/// inputs to a node have changed.
+///
+/// Unlike the RunToken (which is incremented whenever a node re-runs), the Generation is only
+/// incremented when the output of a node has changed.
+///
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct Generation(u32);
+
+impl Generation {
+  fn initial() -> Generation {
+    Generation(0)
+  }
+
+  fn next(&self) -> Generation {
+    Generation(self.0 + 1)
+  }
+}
+
 enum EntryState<N: Node> {
-  NotStarted(RunToken),
+  // A node that has either been explicitly cleared, or has not yet started Running for the first
+  // time. In this state there is no need for a dirty bit because the generation is either in its
```

Comment:
Not to self: the `Generation` is not incremented when a node is cleared: only the RunToken.